src/aosm/azext_aosm/generate_nsd/nsd_generator.py

```diff
@@ -0,0 +1,212 @@
+# --------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT
+# License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------
+"""Contains a class for generating VNF NFDs and associated resources."""
+from knack.log import get_logger
+import json
+import logging
+import os
+import shutil
+from functools import cached_property
+from pathlib import Path
+from typing import Any, Dict, Optional
+
+from azext_aosm.generate_nfd.nfd_generator_base import NFDGenerator
+
+from azext_aosm._configuration import NSConfiguration
+from azext_aosm.util.constants import (
+    NSD_DEFINITION_BICEP_SOURCE_TEMPLATE,
+    NSD_DEFINITION_BICEP_FILE,
+    NF_TEMPLATE_BICEP_FILE,
+    NF_DEFINITION_BICEP_FILE,
+    NSD_ARTIFACT_MANIFEST_BICEP_FILE,
+    NSD_CONFIG_MAPPING_FILE,
+    SCHEMAS,
+    CONFIG_MAPPINGS,
+    NSD_ARTIFACT_MANIFEST_SOURCE_TEMPLATE,
+    TEMPLATES,
+)
+
+from jinja2 import Template
+
+
+logger = get_logger(__name__)
+
+
+class NSDGenerator:
+    """
+    NSD Generator.
+
+    This takes a config file and a set of NFDV deploy_parameters and outputs:
+    - A bicep file for the NSDV
+    - Parameters files that are used by the NSDV bicep file, these are the
+      schemas and the mapping profiles of those schemas parameters
+    - A bicep file for the Artifact manifest
+    - A bicep and JSON file defining the Network Function that will
+      be deployed by the NSDV
+    """
+
+    def __init__(self, config: NSConfiguration):
+        self.config = config
+        self.nsd_bicep_template_name = NSD_DEFINITION_BICEP_SOURCE_TEMPLATE
+        self.nf_bicep_template_name = NF_TEMPLATE_BICEP_FILE
+        self.nsd_bicep_output_name = NSD_DEFINITION_BICEP_FILE
+
+        self.build_folder_name = self.config.build_output_folder_name
+
+    def generate_nsd(self, deploy_parameters) -> None:
+        """Generate a NSD templates which includes an Artifact Manifest, NFDV and NF templates."""
+        logger.info(f"Generate NSD bicep template")
+
+        self.deploy_parameters = deploy_parameters
+
+        self._create_nsd_folder()
+        self.create_parameter_files()
+        self.write_nsd_manifest()
+        self.write_nf_bicep()
+        self.write_nsd_bicep()
+
+        print(f"Generated NSD bicep templates created in {self.build_folder_name}")
+        print(
+            "Please review these templates. When you are happy with them run "
+            "`az aosm nsd publish` with the same arguments."
+        )
+
+    def _create_nsd_folder(self) -> None:
```

Comment:
In cnf and vnf., ive removed this function completely as we now do this if statement in the custom.py itself. Then there is just one line which is creating the folder which you could do in the generate_nsd function. You could do the same?